bokehjs/src/coffee/models/sources/column_data_source.coffee

```diff
@@ -6,20 +6,113 @@ import {SelectionManager} from "../../core/selection_manager"
+#       notice, this list of conditions and the following disclaimer in the
+#       documentation and/or other materials provided with the distribution.
+
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+# ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+# WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+# DISCLAIMED. IN NO EVENT SHALL DANIEL GUERRERO BE LIABLE FOR ANY
+# DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+# (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+# ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+# SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+
+
+#  * Uses the new array typed in javascript to binary base64 encode/decode
+#  * at the moment just decodes a binary base64 encoded
+#  * into either an ArrayBuffer (decodeArrayBuffer)
+#  * or into an Uint8Array (decode)
+#  *
+#  * References:
+#  * https://developer.mozilla.org/en/JavaScript_typed_arrays/ArrayBuffer
+#  * https://developer.mozilla.org/en/JavaScript_typed_arrays/Uint8Array
+
+
+_keyStr = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/="
+
+# will return a  Uint8Array type
+decodeArrayBuffer = (input) ->
+  bytes = parseInt((input.length / 4) * 3, 10)
+  ab = new ArrayBuffer(bytes)
+  decode(input, ab)
+  return ab
+
+removePaddingChars = (input) ->
+  lkey = _keyStr.indexOf(input.charAt(input.length - 1))
+  if (lkey == 64)
+    return input.substring(0,input.length - 1)
+  return input
+
+decode = (input, arrayBuffer) ->
+  # get last chars to see if are valid
+  input = removePaddingChars(input)
+  input = removePaddingChars(input)
+
+  bytes = parseInt((input.length / 4) * 3, 10)
+
+  ab = new ArrayBuffer(bytes)
+  uarray = new Uint8Array(ab);
+
+  input = input.replace(/[^A-Za-z0-9\+\/\=]/g, "")
+
+  i = 0
+  j = 0
+  while i < bytes
+    # get the 3 octects in 4 ascii chars
+    enc1 = _keyStr.indexOf(input.charAt(j++))
+    enc2 = _keyStr.indexOf(input.charAt(j++))
+    enc3 = _keyStr.indexOf(input.charAt(j++))
+    enc4 = _keyStr.indexOf(input.charAt(j++))
+
+    chr1 = (enc1 << 2) | (enc2 >> 4)
+    chr2 = ((enc2 & 15) << 4) | (enc3 >> 2)
+    chr3 = ((enc3 & 3) << 6) | enc4
+
+    uarray[i] = chr1;
+    if enc3 != 64
+      uarray[i+1] = chr2
+
+    if enc4 != 64
+      uarray[i+2] = chr3
+
+    i += 3
+
+  return ab
+
 # Datasource where the data is defined column-wise, i.e. each key in the
 # the data attribute is a column name, and its value is an array of scalars.
 # Each column should be the same length.
 export class ColumnDataSource extends DataSource
   type: 'ColumnDataSource'
 
+  initialize: (options) ->
+    super(options)
+    for k, v of @data
+      bytes = decode(v[0])
```

Comment:
Is there an npm package for this code? We shouldn't be maintaining this ourselves. There are multiple implementations available, e.g. `base64-js` which seems do work exactly as this code (uses typed arrays). Performance is perhaps the only thing to consider.